### ISSUE TITLE
Send review emails to Service contacts

### ISSUE THREAD

**Issue by finnlewis** [localgovdrupal/localgov_workflows#12]
From the Miro board : https://miro.com/app/board/o9J_lHm9M2s=/

@benhillsjones asked about email notifications. 

Can we send notifications to users on workflow changes?

**Comment by stephen-cox** [comment]
@davidsiddall Passing this to you so you can provide the details of what you require here. Once that's defined we can look at specifying this to a point where a developer can build it.

**Comment by davidsiddall** [comment]
As a Contributor, I want to be notified if my edit has been approved or rejected, so that I can decide what to do next (re-edit or start publicising, sharing an approved edit).

As an Editor, I want to automatically notify Contributors after I have reviewed their drafts and approved/rejected accordingly, so that they know what to do next and the content publishing process is as smooth as possible.

At Waltham Forest, we would like automatic email notifications to be sent to those people who have sent their content for approval. 

We have a core team of three content designers (Editors), who will access things sent for approval via the Approval dashboard. At the point of where content transitions from Review to either published or back to draft, we would like an email notification to be triggers.

These emails could use a customisable template. We would also like the revision log to be mandatory when a revision is sent for review or approved/rejected from both the user submitting for review and the Editor approving as per issue #10.

---
Approval email example:

**Subject: Your content (insert page title) has been approved**

(insert username),

Your content (insert page title) has been approved with the following note:

(insert note pulled from the revision log and username of approver)

You can access your content at the following url:

<insert url>

Rejected email example:

**Subject: Your content (insert page title) has been sent back to draft**

(insert username),

Your content (insert page title) has been sent back to draft with the following note:

(insert note pulled from the revision log and username of the approver)

You can continue to edit your content at the following url:

(insert edit link)


---
These notifications could be applied to similar content transitions. For example, when content goes from published to “expired” or “stale” as per issue #3.


@stephen-cox assigning back to you, let me know if you need more details.

**Comment by finnlewis** [comment]
It would be good to confirm if this is something that is needed by at least two councils. 

@benhillsjones @andybroomfield could you review this and comment as to whether this also makes sense for your councils?

Anyone else know of other councils with devolved content management models?

**Comment by willguv** [comment]
@stephen-cox we talked about using David's example to work up a specification for a generic 'create a email notification for a transition from state x to y'

It would be good if we can still do this

As an aside, I spoke to Croydon who don't have a devolved model. @tom-steel and I mused that it might be nice to send an email to a relevant Microsoft Teams channel used by content designers

To do this, we would need the ability to add an optional To: email

I don't think this is MVP, but would be good to discuss and write a card for the future

**Comment by stephen-cox** [comment]
There's a question here about who is notified and how to track this. Drupal only records the author (initial content creator), not the last person to edit a piece of content or the person who changed the moderation state. Although it would be good to confirm this.

@ekes can you think of a way to use revisions to find the person who sent content for revision so a notification can be sent to them?

Otherwise we need to look at adding a 'content owner' field or similar which references a user (or users) to notify when content changes moderation state.

**Comment by finnlewis** [comment]
I think we need a bit more discovery on this to explore the options and flexibility. 

Oxfordshire might be a good council to bring into the discussion... and Lambeth... to confirm mutual requirements and collaborate.

We want this to be modular and optional.

We might need a notifications system to be pluggable: not just emails but other notifications, MS Teams, Slack etc...

**Comment by willguv** [comment]
@finnlewis I agree that we ought to explore this further with other councils and understand more common needs. It's ideal for a reboot of the content lifecycle group at a later date

My slight concern is having nothing might prevent @davidsiddall and Waltham Forest using the remainder of the workflow features. Can you let us know David?

**Comment by davidsiddall** [comment]
@willguv @finnlewis we could run without the email notification and manage it with manual emails for the time being. That way we can also refine the needs based on real life experiences rather than conjecture. But I think it definitely needs a solution in the medium term.

**Comment by andybroomfield** [comment]
I kindas of agree this needs to be some pluggable framework rather than something fixed and hard coded.
Whilst Workflows could provide a best practice suggestions by default, they do need to fit with others content workflows and so be modifiable.

In the Drupal 6/7 days this is something if we where tasked to build we would make use of the [Rules module](https://www.drupal.org/project/rules).
I don't know the state of rules at the moment, however the [Business rules module](https://www.drupal.org/project/business_rules) could be a better fit here than trying to hard code some notifications and we could provide some default rules to acheive this.

**Comment by willguv** [comment]
@stephen-cox please call the tab "Review date" and leave it open by default - thanks

**Comment by andybroomfield** [comment]
This has come up at BHCC, as we are starting to get content appear in review queue since we enabled this module a year ago.
Maybe this is a case to look at [ECA: Event - Condition - Action](https://www.drupal.org/project/ECA), or we could try and provide a basic email template in this module.

**Comment by davidupjohn** [comment]
I'm happy to join in this conversation about the content review part. We have the email notifications at Oxfordshire and the idea of doing it manually or not at all scares me. Agree there needs to be a way to modify the text to explain to the user what the next steps are (might be reply to this email, fill in a form on the intranet or log into drupal and tick something). We don't have a MS Teams for content authors as we don't have any on our public site.

**Comment by GC-92** [comment]
I also support this being added for our staffing needs.

**Comment by willguv** [comment]
Content group thread by David

https://localgovdrupal.slack.com/archives/C020SJRQX2M/p1684492136566859

> how does everyone make sure pages are kept up to date and regularly checked by the page owner/subject matter expert? We have an automated email that sends out a message to the page owner every six months. But they are often ignored

From Stephen

> The main issue with adding this to LGD is making the notifications flexible enough so they can easily be modified for a given situation. There was discussions about being able to set how notifications were sent (email, text, Slack, Teams), but at a minimum email text would need to be adjustable.
> 
> The other issue is how to handle the content ownership side of things. It's fairly easy to add a content owner field to content which will notify one or more users when content is due to review, but it's going to be a pain going through content and adding users to every node and removing them again when they leave or their roles changes.
> 
> With some custom code it would be possible to map content (node IDs) to users (user IDs) in a CSV file or similar and import this into a content owner field. There might already be a module to help with this, but it doesn't come with Drupal.
>

**Comment by willguv** [comment]
I was talking with a council about content ownership this week - public pages, call centre scripts and any published materials. The service owner has a live dashboard but has no idea about calls or web traffic. Has anyone worked this problem from both ends?

**Comment by willguv** [comment]
### Content group chat 23rd May

[write up needs]

Feature list is something like this:

- new authenticated "subject matter expert" user (no publishing permissions)
- assign user(s) to each page. group of pages?
- import spreadsheet of users to get started (email address and nodeIDs, automatically create SMEs)
- send email when review date passed
- BCC copy to web inbox? (acts as an audit log in councils and enables them to check for bouncebacks)

If a new user takes over the same pages as old, can we just update the user?

Creating a new user, could they assume the pages of the old one?

Southwark actively thinking about this. Oxford built something similar years ago for Drupal 7

### Later

- consider integration with Teams, Slack etc
- Haringey: talk to @3mkay about AD integration

**Comment by keelanfh** [comment]
What Essex did on this: https://github.com/essexcountycouncil/content_ownership - could be a starting point for adoption by LGD and future development.

This has:

- Content owners with names and email addresses
- that can be added to content types

Something we know doesn't work too well:

- You have to go to the content ownership page to create new content owners, it would be better to be able to do this in the content itself.

**Comment by willguv** [comment]
@keelanfh has built something minus email notifications - will share

Can't create content owners within a page (have to leave and return)

@stephen-cox issue around changing text/ tokens in messages

@andybroomfield need to decide whether to link to a user account ultimately

Backlog group 24 May are happy

**Comment by davidupjohn** [comment]
In our old Drupal 7 we had a role 'content owner'. Each new user we made we selected that role if they were going to be a content owner. The content owner field is an entity reference to the list of users. Our drupal 9 we've stopped having the role and just add new content owners as authenticated users.

**Comment by willguv** [comment]
@stephen-cox could you install https://github.com/essexcountycouncil/content_ownership on test too please - thanks

**Comment by stephen-cox** [comment]
@willguv I've deployed Essex County Councils Content Ownership module to the test site: https://test.localgovdrupal.org

There's a small bug installing the module which I've reported with a fix here: https://github.com/essexcountycouncil/content_ownership/issues/6

**Comment by davidupjohn** [comment]
> What Essex did on this: https://github.com/essexcountycouncil/content_ownership - could be a starting point for adoption by LGD and future development.
> 
> This has:
> 
> * Content owners with names and email addresses
> * that can be added to content types
> 
> Something we know doesn't work too well:
> 
> * You have to go to the content ownership page to create new content owners, it would be better to be able to do this in the content itself.

OK yes I've tried it. I was expecting it to be under User or a new role/permission. But its a good start. I already have a view we use to list this sort of thing and Agile made us a way to filter. I've no idea how to pass that on to anyone to test ?

**Comment by willguv** [comment]
Discussed at BG today - very useful to Wirral, H&F, Haringey, Brighton

Finn: An efficient way would be to use Oxfordshire as a starting point

This is likely a working group to firm up needs and help test resultant functionality

**Comment by davidupjohn** [comment]
If you want me to run through what we have in Oxfordshire and what works/can be improved let me know.

**Comment by willguv** [comment]
Hi @davidupjohn would love to. Could you demo it at the next session?

**Comment by davidupjohn** [comment]
@willguv Will I'll say 'maybe'. My time is never guaranteed and there's a bit of crappy stuff happening this week with my team. When is the meeting ? I can defo sort out some notes and screen shots so if I can't make it you can get a rough idea and I can answer questions later.

**Comment by willguv** [comment]
Thanks @davidupjohn, here's what you shared with me

https://docs.google.com/document/d/1MOOZ7hKxYeEQlDhs69E2ht6m8Lh8tBdW7KrlmF0TeQ0/edit?usp=sharing

We can talk about it here and at our next meeting on 5th July

**Comment by keelanfh** [comment]
Some connection with this: https://github.com/localgovdrupal/localgov/issues/586 - there could be a case for this "SME" role having access to unpublished content - perhaps in combination with a "Needs review by SME" or similar state that a workflow could move it into after 6 months or something.

**Comment by willguv** [comment]
Andy: 

- send email for a group of pages rather than an email for every page
- run on a set schedule (eg once a month)

@stephen-cox and me to propose an MVP/ estimate and post on Slack

**Comment by davidupjohn** [comment]
That's a basic, low-detail, summary of how it works in Oxfordshire (drupal 9 not LGD) - should be enough for you to get the point across. We can obviously share the code if needs be. It's very old code though and I suspect quite inefficient, but it works after a fashion, so... 

•	We have a small administration form where you define the preferred date and time each month for the review process to run
•	There is a view showing which content is due review, with a contextual filter for user ID
•	The module ties into HOOK_cron in Drupal lingo, every time there is a cron run, Drupal executes all its bits and pieces plus any modules using the cron hook
•	If it's not the right date and time for the content review to happen, nothing happens
•	If it is the right date and time, the view is programmatically loaded for each user appearing in the content due review view, and an email is generated which renders their user-specific version of the view into an email and sends it.

**Comment by davidupjohn** [comment]
"If you build it they will come". So be ready. How will your subject matter expert/page owner tell you a page is OK and that the page review date needs to change? Do you have a web form they fill in? Do they just reply to the email ? And is your web team ready for 100 more emails a month now all the SMEs are engaged with their content.

**Comment by willguv** [comment]
Will and Stephen discussion today. Here's the MVP we're planning to build:

"Service contacts" are entities
- Name of person
- Email address/ Drupal user account name [validate against existing email addresses, accounts]
- "Notes" free text field
- Toggle receive email notifications [on by default]
- Deals with people joining/ leaving (just update the entity)

Assigning ownership
- SC field, add multiple people, a lookup
- Add a “Subject matter expert” widget to sidebar
- [later] add a new SC from this page

SC admin view
- Sort alphabetically, creation date
- List them
- Edit/ Delete/ Add new
- Add to bottom of content menu, and after the “alert banners” content tab

Email alert
- Notify options [send after the review date has expired]
    - Daily/ Weekly/ Monthly on [time/ date]
    - [Later] Send reminder after [x] days
- Edit email message
    - Subject/ Body/ Tokens
    - [Later?] BCC this email address with all notifications



Also adding [agreed 2 August]

Configuration option to enable/ disable email sending

Later 
- Importing spreadsheet
- View of a single subject matter expert that shows the pages they're assigned to?

**Comment by GC-92** [comment]
Just an idea, but in regards to media items, I would be interested to know if notifications and reviews mentioned here for pages, can be added to a media item properties.

Say a PDF was rushed, uploaded on a temporary basis and then forgotten as it was then replaced as HTML following a correct conversion, could there be others out there who think there could be a user need/capability to try and get features like this added to media items?

**Comment by stephen-cox** [comment]
> Just an idea, but in regards to media items, I would be interested to know if notifications and reviews mentioned here for pages, can be added to a media item properties.

The first step would be to get the Review Date module working with all Drupal entities rather than just nodes. This is possible, but would be a bit of work. Once that's working it should be possible to use the notifications we're discussing here.

There's a discussion about extending review date: https://github.com/localgovdrupal/localgov_workflows/issues/51

**Comment by stephen-cox** [comment]
@willguv A very rough estimate of the work required for this is:

- Add subject matter expert (SME) entity extending Essex's work: 3 to 4 hours.
- SME admin view: 2 to 3 hours.
- Create SME widget and add to node pages: 3 to 5 hours.
- Send email when content needs review: 4 to 6 hours.
- Email alert configuration: 2 to 3 hours.

Optional:

- Ability to create new SME entities from node page: 4 to 6 hours
- Send reminder email: 4 to 5 hours.

This comes to 14 to 21 hours for the MVP and 22 to 32 hours with the extra bits.

**Comment by willguv** [comment]
On Joe Bloggs page, can we see Joe Bloggs pages? @stephen-cox

**Comment by willguv** [comment]
Configuration option to enable/ disable email sending

**Comment by keelanfh** [comment]
A few bits to add from our recent exploration of Essex's content ownership feature.

Editors sometimes end up adding duplicate content owners. To prevent this:

* It could be useful to search, filter and sort the list of content owners. The default sort order could be alphabetical rather than in the order that the content owners were created.
* It could be useful to prevent creation of duplicate owners by having a unique constraint on the email address as well as the name, as names can have different variations.

**Comment by willguv** [comment]
Andy: 

- Opt in to emails

**Comment by stephen-cox** [comment]
Came across this which has some interesting ideas about how to managed the notifications: https://www.hashbangcode.com/article/drupal-10-creating-notification-system-using-message-and-eca-module

**Comment by davidupjohn** [comment]
> Came across this which has some interesting ideas about how to managed the notifications: https://www.hashbangcode.com/article/drupal-10-creating-notification-system-using-message-and-eca-module

@stephen-cox 

That link was broken it had an S missing off the end.

https://www.hashbangcode.com/article/drupal-10-creating-notification-system-using-message-and-eca-modules

also 
https://www.codeenigma.com/blog/creating-drupal-processes-event-condition-action-module

Robert and I have looked at that module for OCC but haven't had time to test it yet. Its got huge workflow and process potential

**Comment by willguv** [comment]
@stephen-cox I'm not sure about the term "subject matter expert" so have asked the content group on Slack

My favourite is currently "service contact"

Is it easy to make this value configurable for each council? Thanks

**Comment by stephen-cox** [comment]
> Is it easy to make this value configurable for each council?

It's straightforward to change the displayed names, but is a dev job. Setting this in config is possible, but would be a little awkward.

**Comment by willguv** [comment]
@stephen-cox let's go with Service contact unless you hear from me again before you start - cheers

**Comment by willguv** [comment]
@stephen-cox before starting this work could we also talk about:

- Avoiding Essex having to migrate its existing Service Contacts. Are we building on top of what they've done?
- A view showing all the pages a service contact is responsible for
- searching/ sorting the view of SCs
- clicking on an SC within a page (or rolling over their name?) and seeing contact details 

This is not necessarily MVP, so would be good to chat about it

**Comment by davidupjohn** [comment]
Happy to be involved in this as well. @willguv what were you thinking 'contact details' would be for an SC ? just email address ?

**Comment by willguv** [comment]
Yes think so @davidupjohn

**Comment by willguv** [comment]
@msayoung @agile-simon have you been able to make a start on this? Thanks

**Comment by stephen-cox** [comment]
@andybroomfield @willguv and anyone else who's likely to use review notifications.

I have a question about the type of events that should trigger notification emails.

Should reminder emails be sent whenever content is marked as needing review (i.e. whether manually set or automatically) or just when it's automatically set when a piece of content passes it's needs review date?

From a technical perspective, the question is whether content is added to the notification queue whenever the content moderation field on a piece of content is set to needs review or whether it's added to the queue when a Scheduled Transition sets the content moderation state to needs review.

**Comment by davidupjohn** [comment]
In Oxfordshire we send the email when the content needs review date is reached. I think that is the better way. I don't think all councils will have Scheduled transitions set up.

**Comment by stephen-cox** [comment]
> In Oxfordshire we send the email when the content needs review date is reached. I think that is the better way. I don't think all councils will have Scheduled transitions set up.

The implementation of this in Oxfordshire uses direct SQL queries to find content that has passed it's review date. Having thought about this, I think it's more robust to use a queue which content is added to with specific triggers. It should also make it a bit more extensible if people want to change or add trigger conditions.

Whichever method is used, there will be a dependency on localgov_workflows, which installs Schedule Transitions. The localgov_review_date sub-module configures Schedule Transitions to mark content as needing review when it passes it's review date.

If SQL queries are used to find content needing review then there will be a hard dependency on localgov_review_date as this is where the review date entities are defined, which isn't necessarily the case when using a queue.

**Comment by willguv** [comment]
I think we need to bundle this with Scheduled Transitions (which we need to rename as Scheduling, covered elsewhere)

For now, we should send an email only when the review date is reached. Let's sound out the CDs about doing more at a later date

Thanks Stephen

**Comment by andybroomfield** [comment]
Yes, when review date is reached, we could look at comming to review notifications at a future point.
Given that a lot of content will be hitting review at the same day, it would be better to group them to cut down on email.
Maybe something that could be scheduled so this is what is in the review que, but that could be an advanced feature.
I can talk to comms team about it on what frequency they would like.

**Comment by stephen-cox** [comment]
> Given that a lot of content will be hitting review at the same day, it would be better to group them to cut down on email.

I was planning on adding a way of combining notifications so only one email is sent to a user at a time

> I can talk to comms team about it on what frequency they would like.

This will be configurable. Thinking of defaulting this to daily and having the setting work in days, but could make work in hours if it's likely people will want to use smaller notification periods than a day.

**Comment by willguv** [comment]
Deploy to product site please

**Comment by stephen-cox** [comment]
@willguv The work so far on this has been deployed to the product site. Anyone who wants to test is can install it with:

```shell
composer require localgovdrupal/localgov_workflows:"dev-feature/1.x/12-review-emails as 1.2.3"
drush en localgov_workflows_notifications
```

It's largely working, but needs more tests writing which are likely to uncover bugs. There's a README that provides details on requirements to get this working and a description of the different parts: https://github.com/localgovdrupal/localgov_workflows/blob/feature/1.x/12-review-emails/modules/localgov_workflows_notifications/README.md

Service contacts can be created and managed at /admin/content/localgov-service-contact and there's a widget in the right-hand panel of the node edit form to associate these with content.

The frequency emails are sent at can be set at /admin/config/workflow/localgov-workflows-notifications, it defaults to daily.

The email template can be edited at /admin/config/system/mailer/policy/localgov_workflows_notifications

**Comment by willguv** [comment]
Hi @stephen-cox, I've set up /node/6/edit?destination=/admin/content on the product site to send me a review email tomorrow

Everything looks very clear to use. The only two points I had were:

- could the 'Service contacts' block be higher the sidebar (under the Review block) and be set to open by default
- on /admin/config/workflow/localgov-workflows-notifications there's a value of '0' for email frequency. Does this mean it's sending once a day every day? 

Thanks again for doing this

**Comment by stephen-cox** [comment]
Thanks @willguv 

> could the 'Service contacts' block be higher the sidebar (under the Review block) and be set to open by default

Yep - I'll look at this

> on /admin/config/workflow/localgov-workflows-notifications there's a value of '0' for email frequency. Does this mean it's sending once a day every day?

The default is 1. I just set it to 0 for testing. It will notify for content going into needs review after the last cron run, rather than waiting a day.

**Comment by willguv** [comment]
Hi @stephen-cox the content I set to review is now in the Review tab. I didn't receive an email however.

Was that because the value was set to '0' or is there more to do

Aside from the visibilty of the service contacts block everything else looks great from a product point of view - very clear and simple to use

**Comment by stephen-cox** [comment]
@willguv I have finished the work on this. The latest version is on the product site and you should have just received an email notification from the site.

The PR for the #74 is now ready for a technical review.

**Comment by willguv** [comment]
Provide a view to log the email notifications being sent?

**Comment by willguv** [comment]
Hi @stephen-cox how's the technical review going? Happy to talk about this at today's PG if easier - thanks

**Comment by willguv** [comment]
Blocked on [symfony mailer](https://www.drupal.org/project/symfony_mailer) - how does it play with existing council set up? If they're using default config there will not be any issues. Risk is it may overwrite bespoke work from something like SMTP mailer

**Comment by stephen-cox** [comment]
This feature has been released with LocalGov Workflows 1.3.0: https://github.com/localgovdrupal/localgov_workflows/releases/tag/1.3.0
